File: ios/AppfastflyFingerprint.m
#import "AppfastflyFingerprint.h"
#include <sys/utsname.h>
#include <sys/sysctl.h>

@implementation AppfastflyFingerprint

- (NSDictionary<NSString *, id> *)collect {
  UIDevice *device = [UIDevice currentDevice];
  UIScreen *screen = [UIScreen mainScreen];
  NSLocale *locale = [NSLocale currentLocale];

  NSMutableDictionary *result = [NSMutableDictionary dictionaryWithDictionary:@{
    @"deviceId": [self identifierForVendor],
    @"brand": @"Apple",
    @"model": [self deviceModel],
    @"os": @"iOS",
    @"osVersion": device.systemVersion,
    @"screenWidth": @((NSInteger)(screen.bounds.size.width)),
    @"screenHeight": @((NSInteger)(screen.bounds.size.height)),
    @"screenScale": @(screen.scale),
    @"locale": locale.localeIdentifier,
    @"language": locale.languageCode ?: @"en",
    @"timezone": [NSTimeZone localTimeZone].name,
    @"userAgent": [self defaultUserAgent],
    @"isEmulator": @([self isSimulator]),
  }];

  NSString *country = locale.countryCode;
  if (country) {
    result[@"country"] = country;
  }

  NSString *osBuild = [self osBuildVersion];
  if (osBuild) {
    result[@"osBuild"] = osBuild;
  }

  return [result copy];
}

- (NSString *)identifierForVendor {
  NSUUID *uuid = [UIDevice currentDevice].identifierForVendor;
  return uuid ? uuid.UUIDString : [[NSUUID UUID] UUIDString];
}

- (NSString *)deviceModel {
  struct utsname systemInfo;
  uname(&systemInfo);
  return [NSString stringWithCString:systemInfo.machine encoding:NSUTF8StringEncoding];
}

- (NSString *)osBuildVersion {
  size_t size = 0;
  sysctlbyname("kern.osversion", NULL, &size, NULL, 0);
  if (size == 0) return nil;
  char *build = malloc(size);
  sysctlbyname("kern.osversion", build, &size, NULL, 0);
  NSString *result = [NSString stringWithCString:build encoding:NSUTF8StringEncoding];
  free(build);
  return result;
}

- (NSString *)defaultUserAgent {
  NSString *osVersion = [[UIDevice currentDevice].systemVersion stringByReplacingOccurrencesOfString:@"." withString:@"_"];
  return [NSString stringWithFormat:
    @"Mozilla/5.0 (iPhone; CPU iPhone OS %@ like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148",
    osVersion];
}

- (BOOL)isSimulator {
#if TARGET_OS_SIMULATOR
  return YES;
#else
  return NO;
#endif
}

@end

File: ios/AppfastflyFingerprint.swift
import UIKit
import WebKit

@objc(AppfastflyFingerprint)
public class AppfastflyFingerprint: NSObject {

  @objc public func collect() -> [String: Any] {
    let device = UIDevice.current
    let screen = UIScreen.main
    let locale = Locale.current

    var result: [String: Any] = [
      "deviceId": identifierForVendor(),
      "brand": "Apple",
      "model": deviceModel(),
      "os": "iOS",
      "osVersion": device.systemVersion,
      "screenWidth": Int(screen.bounds.width * screen.scale),
      "screenHeight": Int(screen.bounds.height * screen.scale),
      "screenScale": screen.scale,
      "locale": locale.identifier,
      "language": locale.languageCode ?? "en",
      "timezone": TimeZone.current.identifier,
      "userAgent": defaultUserAgent(),
      "isEmulator": isSimulator(),
    ]

    if let country = locale.regionCode {
      result["country"] = country
    }

    if let osBuild = osBuildVersion() {
      result["osBuild"] = osBuild
    }

    return result
  }

  private func identifierForVendor() -> String {
    return UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
  }

  private func deviceModel() -> String {
    var systemInfo = utsname()
    uname(&systemInfo)
    let machineMirror = Mirror(reflecting: systemInfo.machine)
    let identifier = machineMirror.children.reduce("") { identifier, element in
      guard let value = element.value as? Int8, value != 0 else { return identifier }
      return identifier + String(UnicodeScalar(UInt8(value)))
    }
    return identifier
  }

  private func osBuildVersion() -> String? {
    var size = 0
    sysctlbyname("kern.osversion", nil, &size, nil, 0)
    var build = [CChar](repeating: 0, count: size)
    sysctlbyname("kern.osversion", &build, &size, nil, 0)
    return String(cString: build)
  }

  private func defaultUserAgent() -> String {
    return "Mozilla/5.0 (iPhone; CPU iPhone OS \(UIDevice.current.systemVersion.replacingOccurrences(of: ".", with: "_")) like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148"
  }

  private func isSimulator() -> Bool {
    #if targetEnvironment(simulator)
    return true
    #else
    return false
    #endif
  }
}
